### ISSUE TITLE
Feature: search in DOCX footnotes 

### ISSUE THREAD

**Issue by smfukaya** [dnGrep/dnGrep#966]
Hi!

Please, add support to search in DOCX footnotes (perhaps in header and footer as well).

Thanks in advance!

All the Best!

**Comment by doug24** [comment]
Yes, I will look into doing this.

**Comment by Aliaksandr3by** [comment]
might you add a 'docx to pdf' fast simple converter like pdftopng.exe but doc(x)ToPdf ?

**Comment by doug24** [comment]
@Aliaksandr3by - No. This is well outside of the scope of dnGrep. If it is as fast and simple as you think, then it should be easy to find an application to do this.

**Comment by Aliaksandr3by** [comment]
> @Aliaksandr3by - No. This is well outside of the scope of dnGrep. If it is as fast and simple as you think, then it should be easy to find an application to do this.

oh, sorry, I did think that using a PDF printer is simple. I nothing understand in that. 
![2023-10-21 (1)](https://github.com/dnGrep/dnGrep/assets/11432724/d991b2e4-1231-4554-a7aa-cc07507b0d0b)

**Comment by doug24** [comment]
I'm not going to discuss this other than to say that if you want to print to PDF, you need an application that renders the document file into a display format.  dnGrep does not do that - Microsoft Word does that.  There may be others, but it has nothing to do with dnGrep.

Please do not keep responding off topic in this issue.

**Comment by doug24** [comment]
@smfukaya - I have questions for you, or anyone else interested in footnotes, headers and footers:

When dnGrep extracts text from a docx file, there is no concept of pages: pagination is something done when the document is rendered into display format.  There are a couple of options for footnotes: place them all at the end of the document (like endnotes) or place them after the paragraph containing the footnote reference.  Any preference?  I think after the paragraph is probably a good place.

Second issue is the footnote reference number or mark - should the reference number be inserted into the main text?  It cannot be formatted as a superscript like it would be in the document and including it as a full character may impact searching because it would interrupt the normal word flow if you are doing multiple word searches.

Headers and footers: again, where should they be placed in the plain text output?  My first thought is to put both header and footer at the start of each section where the header or footer changes (but I haven't verified how to do that in code yet).  At a section break they shouldn't interrupt the document text flow.

**Comment by smfukaya** [comment]
Hi @doug24! You have very good points, Sir!

I copied part of an Wikipedia Article ("https://en.wikipedia.org/w/index.php?title=Psyche_(spacecraft)&oldid=1181231343") and save it as a DOCX with some footnotes, header and footer just to make it easier for me see the problem (i didn't insert section breaks though).
[Psyche_(spacecraft).docx](https://github.com/dnGrep/dnGrep/files/13061790/Psyche_.spacecraft.docx)

First, in order to not disrupt too much the document flow, perhaps searching in footnotes could be an option (default unchecked) and searching in header/footers another option (one for both or separately).

Regarding the display results for footnotes, I would prefer to see them as if they were endnotes with their numbers and see the "surrounding" (though there will be many cases that actually they will be separated by many pages) footnotes. And I agreed with you that the main content should not have the reference numbers. The main goal is to determine which document has in the searched string in the footnotes (their connections to the main text could be checked later by opening the documents).

Now regarding headers and footers your suggestion seems good to me while we have some fixed text. However, I added page numbers using fields (the content is the "same" for the whole document footer) in the attached DOCX footer which, if extracted for each page, could cause indeed severe interruption in the document text flow. I think it would make sense to break the document flow only if: 1) it'd be possible to extract the headers/footers for each page of the same section/document whenever there is some kind of variable content (i.e., by using fields); and 2) the user could enable/disable searching in header/footer. I thought this kind of search could be useful to determine which document has a specific string in its header, but I admit its scope is very limited (it can be put aside, if you prefer).

These are my initial thoughts about these issues. If I think of anything else, I'll post it here.

Thank you very much!

All the Best!
